### PR TITLE
add fix for missing xgen pad_token_id

### PR DESCRIPTION
**Pull request by quic-amitraj**
1. Add padding_check_and_fix to handle the case where padding token is out of range of vocab size and not left padded.
2. Removed save_fp32_onnx and save_fp16_onnx from the export.

### CODE REVIEW

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 1]
```diff
@@ -206,6 +206,10 @@ def convert_to_cloud_kvstyle(
 
     if tokenizer.pad_token_id is None:
         tokenizer.pad_token_id = tokenizer.eos_token_id
+        
+    # If Pad token is out of range of vocab size
+    if tokenizer.pad_token_id >= tokenizer.vocab_size:
+        tokenizer.pad_token_id = tokenizer.vocab_size - 1
```
Comment (vbaddi) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:212]: Did the model export and compilation pass through? Outputs verified for model card: `Salesforce/xgen-7b-8k-base`

Comment (quic-amitraj) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:212]: Yes, on local it passed successfully.

Comment (quic-amitraj) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:212]: Let wait for CI to pass and will merge then.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 2]
```diff
@@ -206,6 +206,10 @@ def convert_to_cloud_kvstyle(
 
     if tokenizer.pad_token_id is None:
         tokenizer.pad_token_id = tokenizer.eos_token_id
+        
+    # If Pad token is out of range of vocab size
+    if tokenizer.pad_token_id >= tokenizer.vocab_size:
```
Comment (ochougul) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:211]: Is the `pad_token_id` being assigned on line 208? or it's coming from tokenizer?
If it's coming on line 208, we should move this in above `if` condition.

**Review by ochougul — CHANGES_REQUESTED** [review]

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py (hunk 2 above)
Comment (ochougul) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:211]: makes sense

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 3]
```diff
@@ -205,7 +205,11 @@ def convert_to_cloud_kvstyle(
         tokenizer.padding_side = "left"
 
     if tokenizer.pad_token_id is None:
-        tokenizer.pad_token_id = tokenizer.eos_token_id
+        # If Pad token is out of range of vocab size
+        if tokenizer.pad_token_id >= tokenizer.vocab_size:
+            tokenizer.pad_token_id = tokenizer.vocab_size - 1
```
Comment (vbaddi) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:210]: need to add the change in `load_tokenizer`?

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py (hunk 2 above)
Comment (quic-amitraj) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py]: So should we remove this from export_hf_to_cloud_ai_100 or keep it to support this as a LL API.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py (hunk 3 above)
Comment (quic-amitraj) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:210]: Then, I think there is no need of tokenizer.padding_side != "left" condition as well.

**Review by vbaddi — CHANGES_REQUESTED** [review]

File: QEfficient/utils/_utils.py [hunk 4]
```diff
@@ -128,6 +128,9 @@ def load_hf_tokenizer(model_name: str, cache_dir: Optional[str] = None, hf_token
     model_hf_path = hf_download(repo_id=model_name, cache_dir=cache_dir, allow_patterns=["*.json", "*.py", "*token*"])
     #FIXME(ochougul): should this always return left padded tokenizer?
     tokenizer = AutoTokenizer.from_pretrained(model_hf_path, padding_side=padding_side, trust_remote_code=True, **kwargs)
+    
+    padding_check_and_fix(tokenizer)
```
[the remark this answers is not in the document]
Comment (quic-amitraj) [reply QEfficient/utils/_utils.py:132]: Suggested by @ochougul to handle repetition of code.

[the remark this answers is not in the document]
Comment (vbaddi) [reply QEfficient/utils/_utils.py:132]: there is still repetition of function, we should ideally add check in `load_tokenizer` and use that and raise a `ValueError` in export only. and mention in doc to use `right `padding for LL api users.

[the remark this answers is not in the document]
Comment (ochougul) [reply QEfficient/utils/_utils.py:132]: I think we should check and correct the tokenizer padding if user passes otherwise, as we already have a way to fix it.
So, when we load in our function, we always make sure we pad it the right way and also handle the `eos_token_id`,
while whenever tokenizer is executed at lowest level function, we again check that the tokenizer is correct if not we fix it inside that function.

This handles if user decides to use low level function, it will work irrespective of developer's ignorance.

File: QEfficient/utils/run_utils.py [hunk 5]
```diff
@@ -10,6 +10,7 @@
 import torch
 
 from QEfficient.utils.logging_utils import logger
+from QEfficient.utils._utils import padding_check_and_fix
```
Comment (ochougul) [inline QEfficient/utils/run_utils.py:13]: move to `utils/__init__`

**Review by ochougul — CHANGES_REQUESTED** [review]
Please run linter and rebase.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 6]
```diff
@@ -221,18 +209,15 @@ def convert_to_cloud_kvstyle(
...
 
 def export_kvstyle_transformed_model_to_onnx(model_name: str, transformed_model: torch.nn.Module, tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast],
-                                          onnx_dir_path: str, seq_len: int, save_fp32_onnx: Optional[bool] = False, save_fp16_onnx: Optional[bool] = True):
-    
-    if tokenizer.padding_side != "left":
-        logger.warning("Please use padding_side='left' while initializing the tokenizer")
-        tokenizer.padding_side = "left"
-
-    tokenizer.pad_token_id = tokenizer.eos_token_id if tokenizer.pad_token_id is None else tokenizer.pad_token_id    
+                                          onnx_dir_path: str, seq_len: int, save_fp32_onnx: Optional[bool] = False, save_fp16_onnx: Optional[bool] = True):
```
Comment (anujgupt-github) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:212]: why save fp16 and not fp32?

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 7]
```diff
@@ -465,7 +442,7 @@ def qualcomm_efficient_converter(
 
     """
     # Get model_kv first
-    model_kv = model_kv if model_kv else QEFFCommonLoader.from_pretrained(pretrained_model_name_or_path=model_name, hf_token=hf_token, cache_dir=cache_dir)
+    model_kv = model_kv if model_kv else QEFFCommonLoader.from_pretrained(pretrained_model_name_or_path=model_name, hf_token=hf_token, cache_dir=cache_dir, num_hidden_layers=1)
```
Comment (anujgupt-github) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:445]: whats the reason for this num_hidden_layers=1?

File: QEfficient/utils/__init__.py [hunk 8]
```diff
@@ -13,3 +16,27 @@
...
+    tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast]
+):
+    """
+    Checks Tokenizer paddding side and pad_token_id viability.
+    --------
+    
+    Tokeinizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast]. Pass model tokenizer to check and fix.
+    """
+    
+    if tokenizer.padding_side != "left":
```
Comment (anujgupt-github) [inline QEfficient/utils/__init__.py:32]: can you align with new interface directly?

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
 from QEfficient.src._transformers.auto import QEFFAutoModelForCausalLM
 from QEfficient.src.base import QEFFBaseModel
 from QEfficient.src.common import AUTO_MODEL_MAP_TO_MODEL_TYPE_MAP, QEFF_MODEL_TYPE, QEFFCommonLoader
+import QEfficient.utils
```
Comment (ochougul) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:20]: import properly

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py (hunk 6 above)
Comment (ochougul) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:212]: That's the default behavior to save `fp16_clipped onnx` file, as this is the file used for `qpc` compilation. -> saves storage space.
Whereas we let users save `fp32 onnx` file as well, if he wants to do so via low level APIs

File: QEfficient/utils/__init__.py [hunk 10]
```diff
@@ -13,3 +16,27 @@
     onnx_exists,
     qpc_exists,
 )
+from QEfficient.utils.logging_utils import logger
+
+
+def padding_check_and_fix(
```
Comment (ochougul) [inline QEfficient/utils/__init__.py:22]: move to `_utils.py` file and import here.

File: QEfficient/utils/__init__.py (hunk 8 above)
Comment (ochougul) [reply QEfficient/utils/__init__.py:32]: Hey Anuj, for new interface single change is to change this to `right`, don't want to wait for that PR, we will update this in that PR anyway after rebase.

File: QEfficient/utils/_utils.py [hunk 11]
```diff
@@ -13,6 +13,8 @@
 from requests.exceptions import HTTPError
 from transformers import AutoTokenizer, PreTrainedTokenizer, PreTrainedTokenizerFast
 
+import QEfficient
+import QEfficient.utils
```
Comment (ochougul) [inline QEfficient/utils/_utils.py:17]: import like
`from QEfficient.utils import padding_check_and_fix

**Review by ochougul — CHANGES_REQUESTED** [review]

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py (hunk 6 above)
Comment (quic-amitraj) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:208]: Agree

**Review by ochougul — APPROVED** [review]
LGTM

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py (hunk 6 above)
Comment (quic-amitraj) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:208]: Initially, the default case was that we were saving both the clipped and unclipped versions of the ONNX model, which caused space issues for larger models. However, at the time of compilation, we were using the clipped version of the ONNX model. To handle this situation, we have enabled this feature, where save_fp32 will delete the unclipped version, as its default value is False. It gives user the option to save both, only clipped and only unclipped versions of the ONNX model.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 12]
```diff
@@ -487,9 +410,7 @@ def qualcomm_efficient_converter(
             tokenizer=tokenizer,
             onnx_dir_path=onnx_dir_path,
             seq_length=seq_length,
-            return_path=return_path,
-            save_fp16_onnx=save_fp16_onnx,
-            save_fp32_onnx=save_fp32_onnx)
+            return_path=return_path)
```
Comment (ochougul) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:413]: Is this required now? We can always return path of saved `clipped_fp32` model onnx file?

**Review by ochougul — CHANGES_REQUESTED** [review]

**Review by vbaddi — CHANGES_REQUESTED** [review]
Please remove *`pyc` files before pushing the changing the pr.

**Comment by ochougul** [comment]
Update readme

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 13]
```diff
@@ -43,45 +40,26 @@ def convert_to_cloud_bertstyle(
...
-    if tokenizer.padding_side != "left":
-        logger.warning("Please use padding_side='left' while initializing the tokenizer")
-        tokenizer.padding_side = "left"
-
-    if tokenizer.pad_token_id is None:
-        tokenizer.pad_token_id = tokenizer.eos_token_id
-
+        
     # Decide path for saving exported ONNX files.
-    fp32_model_name, fp16_model_name = export_bertstyle_model_to_onnx(model_name, qeff_model.model, tokenizer, onnx_dir_path, seq_len, save_fp32_onnx, save_fp16_onnx) # type: ignore
```
Comment (ochougul) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:66]: this only returns single string, please correct this.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 14]
```diff
@@ -439,10 +360,7 @@ def qualcomm_efficient_converter(
     hf_token: Optional[str] = None,
     seq_length: int = Constants.seq_length,
     kv: bool = True,
-    return_path: bool = True,
     form_factor: str="cloud",
-    save_fp32_onnx: bool = False,
-    save_fp16_onnx: bool = True,
```
Comment (ochougul) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:445]: make this function return single string as output, which is full path of fp32_clipped onnx file, base_dir is not required to be returned.
Make all functions that are being called underneath this return single string.

File: README.md
Comment (ochougul) [inline README.md:1]: Update LL API usage

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py (hunk 14 above)
Comment (ochougul) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:445]: Make sure wherever this function is being called in whole code-base it expects single string as output from this function.

**Review by ochougul — CHANGES_REQUESTED** [review]
Please add return type strings to functions that you are updating.
Finished review.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py (hunk 13 above)
Comment (quic-amitraj) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:66]: Done

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 15]
```diff
@@ -43,49 +40,26 @@ def convert_to_cloud_bertstyle(
...
-
+    model_name = export_bertstyle_model_to_onnx(model_name, qeff_model.model, tokenizer, onnx_dir_path, seq_len) # type: ignore
+  
     # return the model path for automation.
-    if return_path:
-        if save_fp16_onnx:
-            return onnx_dir_path, os.path.join(onnx_dir_path, f"{fp16_model_name}.onnx")
-        else:
-            return onnx_dir_path, os.path.join(onnx_dir_path, f"{fp32_model_name}.onnx")
+    return onnx_dir_path, os.path.join(onnx_dir_path, f"{model_name}.onnx")
```
Comment (ochougul) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:53]: return single path? no need to return onnx_dir_path

Comment (vbaddi) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:53]: Then this requires changes across this `Api`() usage, check top level api and notebooks, if they need this change.

Comment (vbaddi) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:53]: CI checks are failing

**Review by vbaddi — CHANGES_REQUESTED** [review]

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 16]
```diff
@@ -40,52 +38,31 @@ def convert_to_cloud_bertstyle(
...
 
-def export_bertstyle_model_to_onnx(
-    model_name, model, tokenizer, onnx_dir_path, seq_len, save_fp32_onnx, save_fp16_onnx
-):
+def export_bertstyle_model_to_onnx(model_name, model, tokenizer, onnx_dir_path, seq_len) -> str:
     model_base_name = model_name.replace("/", "_") + "_bertstyle"
     os.makedirs(onnx_dir_path, exist_ok=True)
 
     input_str = Constants.input_str
+    padding_check_and_fix(tokenizer)  # Check and fix tokenizer viability
```
[the remark this answers is not in the document]
Comment (ochougul) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:63]: You are right! will remove from here. But I think the original tokenizer loaded from HF is mostly always right padded, unless we change it.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 17]
```diff
@@ -132,24 +109,22 @@ def export_bertstyle_model_to_onnx(
 
     # Run onnxrt inference
     input_names, ort_outputs = run_model_on_ort(
```
[the remark this answers is not in the document]
Comment (ochougul) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:111]: we will need to provide extra argument for this, we can do this later.

**Review by ochougul — APPROVED** [review]

**Review by vbaddi — APPROVED** [review]
LGTM. Thanks
